app/v1/services/EnrolmentsAuthService.scala

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2020 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1.services
+
+import javax.inject.{Inject, Singleton}
+import play.api.Logger
+import play.api.libs.json.JsResultException
+import uk.gov.hmrc.auth.core.AffinityGroup.{Agent, Individual, Organisation}
+import uk.gov.hmrc.auth.core._
+import uk.gov.hmrc.auth.core.authorise.Predicate
+import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals
+import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals._
+import uk.gov.hmrc.auth.core.retrieve.~
+import uk.gov.hmrc.http.HeaderCarrier
+import v1.models.auth.UserDetails
+import v1.models.errors.{DownstreamError, ForbiddenDownstreamError, LegacyUnauthorisedError}
+import v1.models.outcomes.AuthOutcome
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class EnrolmentsAuthService @Inject()(val connector: AuthConnector) {
+
+  private val authFunction: AuthorisedFunctions = new AuthorisedFunctions {
+    override def authConnector: AuthConnector = connector
+  }
+
+  def authorised(predicate: Predicate)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[AuthOutcome] = {
+    authFunction.authorised(predicate).retrieve(affinityGroup and allEnrolments) {
+      case Some(Individual) ~ _ =>
+        val user: UserDetails =
+          UserDetails(
+            userType = "Individual",
+            agentReferenceNumber = None,
+            clientId = ""
+          )
+
+        successLogging(Individual)
+        Future.successful(Right(user))
+      case Some(Organisation) ~ _ =>
+        val user: UserDetails =
+          UserDetails(
+            userType = "Organisation",
+            agentReferenceNumber = None,
+            clientId = ""
+          )
+
+        successLogging(Organisation)
+        Future.successful(Right(user))
+      case Some(Agent) ~ _ =>
+        val user: UserDetails =
+          UserDetails(
+            userType = "Agent",
+            agentReferenceNumber = None,
+            clientId = ""
+          )
+
+        successLogging(Agent)
+        retrieveAgentReference().map(arn => Right(user.copy(agentReferenceNumber = arn)))
```

Comment:
Most of this can probably be refactored slightly to remove the duplication

Reply:
I made some changes.